riscv/zvk_ext_macros.h

```diff
@@ -0,0 +1,1251 @@
+// "op vd, vs1, vs2", however the resulting output can be confusing when
+// debugging.
+#define ZVK_PRI_REGNUMS_VD_VS2_VS1 \
+  "vd#(%" PRIuREG ") vs2#(%" PRIuREG ") vs1#(%" PRIuREG ")" \
+  " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")"
+#define ZVK_PRV_REGNUMS_VD_VS2_VS1 vd_num, vs2_num, vs1_num, vstart_eg, vl_eg
+
+// Debug format/value macros for instructions taking operands
+// VD, VS2, zimm5. It prints the register numbers of those operands,
+// as well as vstart_eg/vl_eg.
+//
+// This macro is meant to be used with instructions following the template
+// "op vd, vs2, round".
+#define ZVK_PRI_REGNUMS_VD_VS2_ZIMM5 \
+  "vd#(%" PRIuREG ") vs2#(%" PRIuREG ") zimm5(%" PRIuREG ")" \
+  " vstart_eg(%" PRIuREG ") vl_eg(%" PRIuREG ")"
+#define ZVK_PRV_REGNUMS_VD_VS2_ZIMM5 vd_num, vs2_num, zimm5, vstart_eg, vl_eg
+
+// Helper macros to print EGU32x4_t values as 4 uint32_t words,
+// listed in big endian (BE) order (most significant word first),
+// each word being printed in the "natural" order (from top nibble
+// to least nibble), the uint32_t values being printed with the "%08x"
+// format specifier.
+//
+//  printf("This is a 32x4 element group as 4 words: " PRIxEGU32x4 "\n",
+//         PRVEGU32x4_BE(x));
+#define PRIxEGU32x4 "%08x__%08x__%08x__%08x"
+#define PRVEGU32x4_BE(x) \
+    ((x)[3]), ((x)[2]), ((x)[1]), ((x)[0])
+
+// Helper macros to print EGU32x8_t values as 8 uint32_t words,
+// listed in little endian (LE) order (most significant word first),
+// each word being printed in the "natural" order (from top nibble
+// to least nibble), the uint32_t values being printed with the "%08x"
+// format specifier.
+//
+//  printf("This is a 32x8 element group as 8 words: " PRIxEGU32x8 "\n",
+//         PRVEGU32x4_LE(x));
+#define PRIxEGU32x8 "%08x__%08x__%08x__%08x__%08x__%08x__%08x__%08x_"
+#define PRVEGU32x8_LE(x) \
+    ((x)[0]), ((x)[1]), ((x)[2]), ((x)[3]), \
+    ((x)[4]), ((x)[5]), ((x)[6]), ((x)[7])
+
+// Helper macros to print a EGU32x4_t element group, together with
+// the vector number from which is was extracted.
+// The vector number is formatted with "%u", the EGU32x4_t value
+// with the  PRIxEGU32x4 documented above.
+#define PRI_uREG_xEGU32x4 "v%" PRIuREG "(" PRIxEGU32x4 ")"
+#define PRV_REG_EGU32x4_BE(reg_num, reg) (reg_num), PRVEGU32x4_BE(reg)
+#define PRV_REG_EGU32x4_LE(reg_num, reg) (reg_num), PRVEGU32x4_LE(reg)
+
+// Helper macros to print a EGU32x8_t element group, together with
+// the vector number from which is was extracted.
+// The vector number is formatted with "%u", the EGU32x8_t value
+// with the PRIxEGU32x8 documented above.
+#define PRI_uREG_xEGU32x8 "v%" PRIuREG "(" PRIxEGU32x8 ")"
+#define PRV_REG_EGU32x8_LE(reg_num, reg) (reg_num), PRVEGU32x8_LE(reg)
+
+//
+// Predicate Macros
+//
+
+// Ensures that the ZVBB extension (vector crypto bitmanip) is present,
+// which depends on vector.
+#define require_zvbb \
+  do { \
+    require_vector(true); \
+    require_extension(EXT_ZVBB); \
+  } while (0)
```

Comment:
If Zvbb requires V, that should be enforced in `isa_parser.cc`. It does not need to be enforced here in every instruction. (Same comment for all of this family of macros; it looks like none of these macros are needed.)

Reply:
I think the point here is to raise an exception when an instruction is executed without proper support. I am not that familiar with spike, but I would imagine `require_extension(EXT_ZVBB);` is required to raise the exception if the instruction is not supported. The dependency on V (which may be V or Zve* I think) should indeed be handled when the ISA support is initialized.

Reply:
The `require_vector(true)` checks `sstatus.VS` and `vtype.vill`.  So, even though the ISA parser should enforce the extension dependency, this code (or something similar) is still necessary.

Reply:
> The `require_vector(true)` checks `sstatus.VS` and `vtype.vill`. So, even though the ISA parser should enforce the extension dependency, this code (or something similar) is still necessary.

Good point. Should those checks be part of `require_extension(EXT_ZVBB)`, do you think?

Reply:
IMO, the current formulation is OK.

The only problem with it, I think, is that it assumes Zvbb requires V.  This isn't strictly true; Zvbb only requires a subset of V: Zve64x, IIRC.  But since Spike doesn't support any of the V subsets currently, that is not a problem currently.

Furthermore, if Spike is augmented to support Zve64x etc., then essentially all of the vector instructions will need to be retooled in the same manner as this one.  In other words, this approach doesn't create a new type of technical debt.

Reply:
Thanks for the discussion. I updated the documentation of those new require_xyz macros to state 
```
// Ensures that the ... extension is present,
// and the vector unit is enabled and in a valid state.```

This captures the semantics of require_vector(true) more clearly than the previous "and that vector is present".